### ISSUE TITLE
Is keylime adding PCR 16 to the quote?  Is it byte flipping the PCR selection?

### ISSUE THREAD

**Issue by kgold2** [keylime/keylime#613]
Is your an issue a feature request? If so, please raise it as an [enhancement](https://github.com/keylime/enhancements)

# Environment
 * OS / version: fedora 32
 * Processor architecture: x86
 * TPM Manufacturer: sw
 * Keylime version: master

# Description 

1. The keylime.conf says that tpm_policy is PCR 22, 15

-  Is this what keylime is quoting?  
- What are those long strings like 000000... 0001?  They appear hash'y, but there are no algorithms.  What would happen if there are two hash algorithms with the same length?
- Since the largest length is 96, this seems line SHA-384 but the text says SHA-512.
- Why does 22 have 6 entries, while 15 has 3 entries?

2. When keylime quotes, the mask is 408000.  Is that the PCR selection, but byte flipped from the TPM format?

3.When I probe the quote with wireshark, I see 0b 03 00 80 41.  0b is sha256, 03 is a 3-byte selection.  However 008041 is PCR 15, 16, and 22.  Why the difference?  Why is PCR 16 being quoted?

4. With wireshark, I also see a pcrreset and then pcrextend to PCR 16.  What's that for?

5. After the quote, I see a read of PCR 15, 16, 22.   Why PCR 16?

7. Another process can extend PCR 16 (the debug PCR) between the quote and the pcrread, so it is unreliable.

## Expected behavior vs. actual behavior 

If the configuration file says 15 and 22, I would not expect 16 to be quoted.

I didn't expect the GET byte flip, if that's what's happening.

The byte little endian PCR selection is used for backward compatibility.  For example, a quote of PCR 0-15 can be 3 bytes ff ff 00 or 2 bytes ff ff.  If a future TPM has more than 32 PCRs, it will interpret the selection correctly.

## Steps to reproduce problem
1. Run keylime
2. Capture packets with wireshark

# Relevant logs

**Comment by nabilschear** [comment]
yes, keylime automatically includes re-settable PCR 16 in all quote responses to bind the NK public transport key to the quote.  See area B of figure 5 in https://github.com/keylime/keylime/blob/master/doc/tci-acm.pdf

**Comment by stefanberger** [comment]
>     1. When keylime quotes, the mask is 408000.  Is that the PCR selection, but byte flipped from the TPM format?

It looks like this would be big endian/human readable format 0x408000 and thus bits 15 and 22 are set. So this should be correct.

> 3.When I probe the quote with wireshark, I see 0b 03 00 80 41. 0b is sha256, 03 is a 3-byte selection. However 008041 is PCR 15, 16, and 22. Why the difference? Why is PCR 16 being quoted?

This PCR selection bit mask written in big endian format would be 0x418000. It looks like bit 16 was added, but otherwise it's a transformation of 0x408000.

**Comment by nabilschear** [comment]
here's the code were PCR 16 is added: https://github.com/keylime/keylime/blob/cbf963b77763587182f80c3bb183af7087d8da63/keylime/tpm/tpm_main.py#L936-L944

**Comment by nabilschear** [comment]
arguably, we should move that mask check into the `if data is not None:` block, but it's still ok

**Comment by kgold2** [comment]
I suggest that using PCR 16 will be unreliable.  NK seems to be the U/V application from the paper.  PCR 16 is is reserved as the 'debug PCR'.  Many applications will use it. A reset or extend during the keylime quote process causes the quote to fail and the tenant to stop.
With keylime quoting every second, the failure is easy to reproduce.
This could either be unintentional - an application using PCR 16 to debug - or malicious - a DoS attack on keylime.

**Comment by nabilschear** [comment]
is there an unused resettable PCR that we can use instead?  I just picked the first resettable one.

**Comment by kgold2** [comment]
Not today.  

A bit of politics:  I have been trying for a TCG 'registry' of PCR usage.  If there other TCG member companies viewing this, please encourage your representatives to support a PCR registry.

Until then, is it correct that the NK and PCR 16 are not related to attestation, but the U/V in the paper is a separate application?  If so, can it be removed from the base keylime product?

If it is essential to attestation, the typical solution is an TPM NV 'extend' index, which acts like a PCR but has authorization.  There is a TCG registry where an NV index can be reserved.  Performance should be better because keylime currently does the PCR reset/extend on every attestation, while NV perhaps only has to be extended once or once per boot.

TPM NV is also limited, but it's not as scarce as PCRs, and they can be reserved.

**Comment by nabilschear** [comment]
the hash of the public NK needs to be bound to the quote as part of delivering the U/V keys.  We used to just hash it with the nonce and keep it there.  But there was some MITM attack we wanted to avoid (see slide 18 of https://www.opensecuritytraining.info/IntroToTrustedComputing_files/Day2-1-auth-and-att.pdf )  So, we picked a resettable PCR.  16 happened to be the first one.  This can be changed to any resettable PCR.  cc @lkatalin related to our conversation today about NK

**Comment by kgold2** [comment]
My question is really not '**which** PCR to use' - for which the answer is that none is guaranteed to work.

Is the U/V keys part of keylime **essential** to the remote attestation solution?  My sense is that it it not - it's a separate application.  If so, I suggest that it be hoisted out of the core keylime solution.  As a separate application, the AK can sign other data.  For example, the NK public key could be directly hashed and signed.  It could be extended into an NV index and certified.

The problem with PCR 16 is that any application can extend it and immediately break attestation.

**Comment by THS-on** [comment]
This can be closed in favor of https://github.com/keylime/enhancements/issues/60 which includes a way to eliminate the usage of PCR16 by using TPM2_Certify. 

cc: @mpeters
